### PR TITLE
#24283: Add DP Support for YoloV4

### PR DESCRIPTION
**Pull request by kkeerthana0573**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/24283)

### Problem description
Add DataParallel support for Yolov4 model. 

### What's changed
Added data parallel support for Yolov4 model to :
- Demo,
- e2e performant.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [x] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by kkeerthana0573** [comment]
CIs passing links (02.07.2025): 
- [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/16004055588) - [N300](https://github.com/tenstorrent/tt-metal/actions/runs/16004055588/job/45146775298#step:5:292)
- [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/16004060034) - [N150](https://github.com/tenstorrent/tt-metal/actions/runs/16004060034/job/45146716638#step:9:412), [N300](https://github.com/tenstorrent/tt-metal/actions/runs/16004060034/job/45146716764#step:9:420)
- [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/16004073478/job/45146908955#step:8:1015) 
- [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/16004063449/job/45146783166#step:5:2887)
- [(Single-card) Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/runs/16004066948) - [N150](https://github.com/tenstorrent/tt-metal/actions/runs/16004066948/job/45146752605#step:8:484) - Passed, [N300](https://github.com/tenstorrent/tt-metal/actions/runs/16004066948/job/45146752692) - Setup failure.

**Review by dvartaniansTT — APPROVED** [review]

**Comment by dvartaniansTT** [comment]
@kkeerthana0573 please resolve the conflicts

File: models/demos/yolov4/demo.py [hunk 1]
```diff
@@ -2,24 +2,178 @@
...
+    torch_input_tensor = torch.cat(torch_images, dim=0)
+    tt_output = runner.run(torch_input_tensor)
+
+    conf_thresh = 0.3
+    nms_thresh = 0.4
+    boxes = post_processing(torch_images, conf_thresh, nms_thresh, tt_output)
+
+    namesfile = "models/demos/yolov4/resources/coco.names"
+    class_names = load_class_names(namesfile)
+    img_cv = cv2.imread(paths_images[0])
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: This seems to be unused

File: models/demos/yolov4/demo.py [hunk 2]
```diff
@@ -2,24 +2,178 @@
...
+    tt_output = runner.run(torch_input_tensor)
+
+    conf_thresh = 0.3
+    nms_thresh = 0.4
+    boxes = post_processing(torch_images, conf_thresh, nms_thresh, tt_output)
+
+    namesfile = "models/demos/yolov4/resources/coco.names"
+    class_names = load_class_names(namesfile)
+
+    img_cv = cv2.imread(paths_images[0])
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: Here as well

File: models/demos/yolov4/demo.py [hunk 3]
```diff
@@ -2,24 +2,178 @@
...
+    with open(os.path.expanduser("~") + "/fiftyone/coco-2017/info.json", "r") as file:
+        coco_info = json.load(file)
+        class_names = coco_info["classes"]
+
+    # Preprocess batch of images to [BS, C, H, W]
+    torch_images = []
+    orig_images = []
+    paths_images = []
+    for batch in data_set:
+        paths, im0s, s = batch
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: S seems to be unused var

File: models/demos/yolov4/demo.py [hunk 4]
```diff
@@ -2,24 +2,178 @@
...
+        batch_size_per_device,
+        act_dtype,
+        weight_dtype,
+        resolution=resolution,
+        model_location_generator=model_location_generator,
+        mesh_mapper=inputs_mesh_mapper,
+        mesh_composer=output_mesh_composer,
+    )
+
+    image_files = [os.path.join(input_loc, f) for f in os.listdir(input_loc) if f.endswith((".jpg", ".png"))]
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: This will not catch .JPG or .PNG cases, consider using lower()

File: models/demos/yolov4/demo.py [hunk 5]
```diff
@@ -2,24 +2,178 @@
...
+    orig_images = []
+    paths_images = []
+
+    for i in range(batch_size):
+        image_path = image_files[i]
+        image = cv2.imread(image_path)
+        if image is None:
+            raise ValueError(f"Image not found or unreadable: {image_path}")
+
+        orig_images.append(image.copy())
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: Do we need to copy images here? Are they being modified?

File: models/demos/yolov4/demo.py [hunk 6]
```diff
@@ -2,24 +2,178 @@
...
+        act_dtype,
+        weight_dtype,
+        resolution=resolution,
+        model_location_generator=model_location_generator,
+        mesh_mapper=inputs_mesh_mapper,
+        mesh_composer=output_mesh_composer,
+    )
+
+    dataset = fiftyone.zoo.load_zoo_dataset("coco-2017", split="validation", max_samples=batch_size)
+    sampled = random.sample(list(dataset), batch_size)
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: Is there a reason we want this random not to be seeded?

File: models/demos/yolov4/demo.py [hunk 7]
```diff
@@ -2,24 +2,178 @@
...
+        if image is None:
+            raise ValueError(f"Image not found or unreadable: {image_path}")
+
+        orig_images.append(image.copy())
+        paths_images.append(image_path)
+        tensor = preprocess_image(image, resolution)
+        torch_images.append(tensor.unsqueeze(0))
+
+    torch_input_tensor = torch.cat(torch_images, dim=0)
+    tt_output = runner.run(torch_input_tensor)
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py:76]: Is it useful to add some kind of output validation at the end?

File: models/demos/yolov4/tests/perf/test_e2e_performant.py [hunk 8]
```diff
@@ -9,58 +9,118 @@
...
-
-    inference_time_avg = round(sum(inference_times) / len(inference_times), 6)
+    inference_time_avg = sum(inference_times) / len(inference_times)
     logger.info(
-        f"average inference time (ms): {inference_time_avg * 1000}, average throughput (fps): {round(batch_size/inference_time_avg)}"
+        f"Model: ttnn_yolov4 - batch_size: {batch_size}, resolution: {resolution}. One inference iteration time (sec): {inference_time_avg}, FPS: {(batch_size * num_devices)/inference_time_avg}"
     )
 
-    expected_inference_time = batch_size / expected_inference_throughput
-    prep_perf_report(
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/tests/perf/test_e2e_performant.py]: Why did we remove this prep_per_report it is a recent addition by @esmalTT

File: models/demos/yolov4/demo.py (hunk 1 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed.

File: models/demos/yolov4/demo.py (hunk 2 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed.

File: models/demos/yolov4/demo.py (hunk 3 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Addressed.

File: models/demos/yolov4/demo.py (hunk 4 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Addressed.

File: models/demos/yolov4/demo.py (hunk 6 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Added reset_seeds.

File: models/demos/yolov4/demo.py (hunk 7 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py:76]: We do validate in _execute_yolov4_trace_2cqs_inference which is in runner.run.

File: models/demos/yolov4/tests/perf/test_e2e_performant.py (hunk 8 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/tests/perf/test_e2e_performant.py]: PR was brought up few days back. I added perp_perf_report now.

File: models/demos/yolov4/demo.py (hunk 5 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed copy.

File: models/demos/yolov4/demo.py (hunk 6 above)
Comment (dstoiljkovicTT) [reply models/demos/yolov4/demo.py]: I can see that you added this variable, not sure of the naming but that's not the main issue, reset_seeds is still unused, random is not seeded you are missing the call random.seed(num)

File: models/demos/yolov4/demo.py (hunk 7 above)
Comment (dstoiljkovicTT) [reply models/demos/yolov4/demo.py:76]: Yes thank you, I can see that in runner.run right after _execute_yolov4_trace_2cqs_inference, _validate is called but only if flag check_with_pcc is set which defaults to false. Not saying you should change it, just trying to understand if we are not doing the validation of the output why not

Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py:76]: In the demo, we currently focus on generating output predictions and do not validate tt_outputs against torch_outputs. Would you suggest enabling check_pcc?

Comment (dstoiljkovicTT) [reply models/demos/yolov4/demo.py:76]: No need to enable it, if that's your decision to keep it that way, but maybe just for your knowledge, for you to know where the demo stands, have you checked what is the pcc on this?

File: models/demos/yolov4/demo.py (hunk 6 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: removed random.sample(..) as it was redundant.

Comment (dstoiljkovicTT) [reply models/demos/yolov4/demo.py]: Great, thank you, could you remove the unused reset_seeds you just added?

File: models/demos/yolov4/demo.py (hunk 7 above)
Comment (ctr-rsaravananTT) [reply models/demos/yolov4/demo.py:76]: PCC is 0.99 for DP and also 0.99 for single chip. Getting FPS as mentioned in Readme.md for both single and Multi run

File: models/demos/yolov4/demo.py [hunk 9]
```diff
@@ -2,24 +2,209 @@
...
+)
+@pytest.mark.parametrize(
+    "resolution",
+    [
+        (320, 320),
+        (640, 640),
+    ],
+)
+def test_yolov4(
+    device, input_loc, batch_size_per_device, act_dtype, weight_dtype, model_location_generator, resolution, reset_seeds
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: Could you remove reset_seeds you added with this PR since it is not used or explain its functionality?

File: models/demos/yolov4/demo.py [hunk 10]
```diff
@@ -29,54 +214,90 @@
...
 )
+def test_yolov4_dp(
+    mesh_device,
+    input_loc,
+    batch_size_per_device,
+    act_dtype,
+    weight_dtype,
+    model_location_generator,
+    resolution,
+    reset_seeds,
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: Here again remove reset_seeds you added with this PR since it is not used or explain what do we use it for?

File: models/demos/yolov4/demo.py [hunk 11]
```diff
@@ -29,54 +214,90 @@
...
+def test_yolov4_coco(
     device,
-    batch_size,
+    batch_size_per_device,
     act_dtype,
     weight_dtype,
     model_location_generator,
     resolution,
-    imgfile,
+    reset_seeds,
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: Could you remove reset_seeds you added with this PR since it is not used?

File: models/demos/yolov4/demo.py [hunk 12]
```diff
@@ -29,54 +214,90 @@
...
+    ],
+)
+def test_yolov4_coco_dp(
+    mesh_device,
+    batch_size_per_device,
+    act_dtype,
+    weight_dtype,
+    model_location_generator,
+    resolution,
+    reset_seeds,
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/demo.py]: Here as well

File: models/demos/yolov4/tests/pcc/test_ttnn_yolov4.py [hunk 13]
```diff
@@ -40,7 +40,7 @@ def run_yolov4(device, reset_seeds, model_location_generator, use_pretrained_wei
         torch_model.load_state_dict(new_state_dict)
         torch_model.eval()
 
-    imgfile = "models/demos/yolov4/resources/giraffe_320.jpg"
+    imgfile = "models/demos/yolov4/resources/giraffe.jpg"
```
Comment (dstoiljkovicTT) [inline models/demos/yolov4/tests/pcc/test_ttnn_yolov4.py:43]: There is one unused variable reset_seeds in definition of the run_yolo4 in this files line 28 could you please either remove it or explain its function?

Comment (kkeerthana0573) [reply models/demos/yolov4/tests/pcc/test_ttnn_yolov4.py:43]: I hadn't brought up the test earlier. as there was a naming conflict with input_image, I needed to resolve it. I've now addressed and removed reset_seeds as discussed.

File: models/demos/yolov4/demo.py (hunk 12 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed reset_seeds and Updated.

File: models/demos/yolov4/demo.py (hunk 11 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed reset_seeds and Updated.

File: models/demos/yolov4/demo.py (hunk 10 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed reset_seeds and Updated.

File: models/demos/yolov4/demo.py (hunk 9 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed reset_seeds and Updated.

File: models/demos/yolov4/demo.py (hunk 6 above)
Comment (kkeerthana0573) [reply models/demos/yolov4/demo.py]: Removed reset_seeds and Updated.

File: models/demos/yolov4/tests/pcc/test_ttnn_yolov4.py (hunk 13 above)
Comment (dstoiljkovicTT) [reply models/demos/yolov4/tests/pcc/test_ttnn_yolov4.py:43]: Great thank you, just make sure not to vreak blackhole testing of yolov4 in models/demos/yolov4/tests/pcc/test_ttnn_yolov4_bh.py, you forgot to erase the arg from the list there

**Review by dstoiljkovicTT — APPROVED** [review]
Make sure to take into consideration and correct last comment I added regarding bh testing, apart from that this looks good to me

**Review by ctr-rsaravananTT — APPROVED** [review]
